lib/libuzfs/libuzfs_attr.c

```diff
@@ -0,0 +1,690 @@
+		return (err);
+	}
+
+	dmu_tx_t *tx = dmu_tx_create(os);
+	dmu_tx_hold_sa(tx, sa_hdl, B_FALSE);
+	if ((err = dmu_tx_assign(tx, TXG_WAIT)) != 0) {
+		dmu_tx_abort(tx);
+		return (err);
+	}
+
+	sa_bulk_attr_t sa_attrs[UZFS_END];
+	int cnt = 0;
+	libuzfs_add_bulk_attr(dhp, sa_attrs, &cnt, (uzfs_attr_t *)attr);
+	VERIFY0(sa_bulk_update(sa_hdl, sa_attrs, cnt, tx));
+	sa_handle_destroy(sa_hdl);
+	if (txg != NULL) {
+		*txg = tx->tx_txg;
+	}
+	dmu_tx_commit(tx);
+	return (0);
+}
+
+static int
+libuzfs_get_nvlist_from_handle(const sa_attr_type_t *sa_tbl,
+    nvlist_t **nvl, sa_handle_t *sa_hdl)
+{
+	int xattr_size;
+	int err = sa_size(sa_hdl, sa_tbl[UZFS_XATTR], &xattr_size);
+	if (err != 0) {
+		return (err);
+	}
+
+	char *obj = vmem_alloc(xattr_size, KM_SLEEP);
+	err = sa_lookup(sa_hdl, sa_tbl[UZFS_XATTR], obj, xattr_size);
+	if (err != 0) {
+		return (err);
+	}
+
+	err = nvlist_unpack(obj, xattr_size, nvl, KM_SLEEP);
+	vmem_free(obj, xattr_size);
+	return (err);
+}
+
+static int
+libuzfs_kvattr_set_nvlist(nvlist_t *nvl, int err_check_sa,
+    int *err_check_sa_enough, const char *name,
+    const char *value, size_t value_size, size_t *xattr_sa_size)
+{
+	int err = 0;
+	if (value == NULL && err_check_sa == 0) {
+		err = nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
+	} else if (value != NULL) {
+		err = nvlist_add_byte_array(nvl, name,
+		    (uchar_t *)value, (uint_t)value_size);
+	}
+	if (err != 0) {
+		return (err);
+	}
+
+	err = nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR);
+	if (err != 0) {
+		return (err);
+	}
+
+	if (*xattr_sa_size > DXATTR_MAX_SA_SIZE ||
+	    *xattr_sa_size > SA_ATTR_MAX_LEN) {
+		*err_check_sa_enough = EFBIG;
+		err = nvlist_remove(nvl, name, DATA_TYPE_BYTE_ARRAY);
+		if (err != 0 && value != NULL) {
+			return (err);
+		}
+		return (nvlist_size(nvl, xattr_sa_size, NV_ENCODE_XDR));
+	}
+	return (err);
+}
+
+static void
+libuzfs_save_xattr_dir(sa_handle_t *sa_hdl, libuzfs_dataset_handle_t *dhp,
+    uint64_t xattr_obj, uint64_t value_obj, const char *name,
+    const char *value, size_t size, dmu_tx_t *tx)
+{
+	sa_attr_type_t *sa_tbl = dhp->uzfs_attr_table;
+	objset_t *os = sa_hdl->sa_os;
+	int dnodesize = dmu_objset_dnodesize(os);
+	if (xattr_obj == DMU_NEW_OBJECT) {
+		xattr_obj = zap_create_dnsize(os, DMU_OT_DIRECTORY_CONTENTS,
+		    DMU_OT_NONE, 0, dnodesize, tx);
+		VERIFY0(sa_update(sa_hdl, sa_tbl[UZFS_DXATTR],
+		    &xattr_obj, 8, tx));
+	}
+
+	if (value_obj == DMU_NEW_OBJECT) {
+		libuzfs_create_inode_with_type_impl(dhp, &value_obj,
+		    B_FALSE, INODE_FILE, tx);
+		VERIFY0(zap_update(os, xattr_obj, name,
+		    8, 1, &value_obj, tx));
+	}
+
+	dmu_write(os, value_obj, 0, size, value, tx);
+
+	sa_handle_t *value_obj_hdl = NULL;
+	VERIFY0(sa_handle_get(os, value_obj, NULL,
+	    SA_HDL_PRIVATE, &value_obj_hdl));
+	VERIFY0(sa_update(value_obj_hdl,
+	    sa_tbl[UZFS_SIZE], &size, 8, tx));
+	sa_handle_destroy(value_obj_hdl);
+}
+
+int
+libuzfs_inode_set_kvattr(libuzfs_dataset_handle_t *dhp, uint64_t ino,
```

Comment:
It looks to me the whole process is like this:

1. check if the name exist in sa (exist_in_sa) and check if sa is big enough for new value (store_in_sa)
    - 1.1 if exist_in_sa
         - 1.1.1 if store_in_sa , do store in sa and return (likely)
2. check if the name exsit in dir (exsit_in_dir),
    - 2.1. if exist_in_dir
         - 2.1.1  if store_in_sa , do store in sa and delete obj in dir and return
         - 2.1.2 else, update obj in dir and return
   - 2.2. if not exist_in_dir (likely)
        - 2.2.1 if store_in_sa, do store in sa and return (likely)
        - 2.2.2 else, create obj in dir and return

The key point here is, it's unlikely to have dir obj, thus 1.1.1 and 2.2.1 are frequent path, we could avoid do step 2 if 1.1.1 is true.

In additition, these err_check_* looks a little bit complicated, and their meanings are obscure, could we simplify it by introducing some easy-to-understand flags like: exist_in_sa/store_in_sa?